src/GoTrueClient.ts

```diff
@@ -1060,42 +1064,91 @@ export default class GoTrueClient {
+   * platform's foreground indication mechanism and call these methods
+   * appropriately to conserve resources.
+   *
+   * {@see #stopAutoRefresh}
+   */
+  async startAutoRefresh() {
+    await this.stopAutoRefresh()
+    this.autoRefreshTicker = setInterval(
+      () => this._autoRefreshTokenTick(),
+      AUTO_REFRESH_TICK_DURATION
+    )
+
+    // run the tick immediately
+    await this._autoRefreshTokenTick()
+  }
+
+  /**
+   * Stops an active auto refresh process running in the background (if any).
+   * See {@link #startAutoRefresh} for more details.
+   */
+  async stopAutoRefresh() {
+    const ticker = this.autoRefreshTicker
+    this.autoRefreshTicker = null
+
+    if (ticker) {
+      clearInterval(ticker)
     }
   }
 
   /**
-   * Clear and re-create refresh token timer
-   * @param value time intervals in milliseconds.
-   * @param session The current session.
+   * Runs the auto refresh token tick.
    */
-  private _startAutoRefreshToken(value: number) {
-    if (this.refreshTokenTimer) clearTimeout(this.refreshTokenTimer)
-    if (value <= 0 || !this.autoRefreshToken) return
+  private async _autoRefreshTokenTick() {
+    const now = Date.now()
 
-    this.refreshTokenTimer = setTimeout(async () => {
-      this.networkRetries++
+    try {
       const {
         data: { session },
-        error: sessionError,
+        error,
       } = await this.getSession()
-      if (!sessionError && session) {
-        const { error } = await this._callRefreshToken(session.refresh_token)
-        if (!error) this.networkRetries = 0
-        if (
-          error instanceof AuthRetryableFetchError &&
-          this.networkRetries < NETWORK_FAILURE.MAX_RETRIES
-        )
-          this._startAutoRefreshToken(NETWORK_FAILURE.RETRY_INTERVAL ** this.networkRetries * 100) // exponential backoff
+
+      if (!session || !session.refresh_token || !session.expires_at) {
+        return
       }
-    }, value)
-    if (typeof this.refreshTokenTimer.unref === 'function') this.refreshTokenTimer.unref()
+
+      // session will expire in this many ticks (or has already expired if <= 0)
+      const expiresInTicks = Math.floor((now - session.expires_at) / AUTO_REFRESH_TICK_DURATION)
+
+      if (expiresInTicks < AUTO_REFRESH_TICK_THRESHOLD) {
+        await this._callRefreshToken(session.refresh_token)
+      }
+    } catch (e: any) {
+      console.error('Auto refresh tick failed with error. This is likely a transient error.', e)
+    }
   }
 
+  /**
+   * Registers callbacks on the browser / platform, which in-turn run
+   * algorithms when the browser window/tab are in foreground. On non-browser
+   * platforms it assumes always foreground.
+   */
   private _handleVisibilityChange() {
     if (!isBrowser() || !window?.addEventListener) {
+      if (this.autoRefreshToken) {
+        // in non-browser environments the refresh token ticker runs always
+        this.startAutoRefresh()
+      }
+
```

Comment:
why is there a need to auto refresh if we're recommending the user to create a client for each api route handler on the server-side and not a globally shared one?

Reply:
This is actually for Flutter, React Native or other types of non-browser (but also non-server) environments.

Reply:
On server-side you don't ever want a ticker like this to be registered. It likely won't run or will run at the worst possible time. We need to make that explicit in docs.

Reply:
Slightly confused here - would users in a server-side environment have to know that they have to set the `autoRefreshToken` option to false  so that the ticker doesn't run

seems like the `!isBrowser` check is `!(typeof window !== 'undefined')`  which looks like it would  be true for server side environments

Reply:
@kangmingtay @hf,

My apologies for being late to the party here. I read this as @J0 does - that this would also match a server client. And, because we don't have clear docs that server-side clients should set `autoRefreshToken` to `false`, would firing off the `startAutoRefresh` code cause app issues where a server-side client exists with `autoRefreshToken` not explicitly set to `false`?

Not to cross-contaminate this, but it's a close sibling to the issue Gary created [here](https://github.com/supabase/gotrue-js/issues/539); where we need better docs on some of these auth options for a server-side client. Gary would rather throw an error to handle it, but I'm not sure if that's warranted unless there are negative consequences to an app running in these scenarios.